### ISSUE TITLE
Add annotation to pass region to vault agent

### ISSUE THREAD

**Issue by infa-mhadiman** [hashicorp/vault-k8s#244]
**Is your feature request related to a problem? Please describe.**
The agent sidecar injected fails to perform an auth on gov-cloud env’s. A **SignatureDoesNotMatch** error is seen, agent uses the global sts endpoint (us-east-1) by default but for gov clouds, we have to use the regional endpoints.

Method – aws
Type - IAM
Vault - 1.5

**Describe the solution you'd like**
Please add an annotation which adds the region to the agent config.

**Comment by jasonodonnell** [comment]
Hi @infa-mhadiman, thanks for the feature request!

Sorry for my ignorance, but this might already be supported using `vault.hashicorp.com/auth-config-region: us-west-2`:

`auth-config` allows you to specify any additional parameters you want to set on your auth method, so this should work in theory. Can you give it a shot?

**Comment by infa-mhadiman** [comment]
![image](https://user-images.githubusercontent.com/54311870/116513192-d423ea00-a8e6-11eb-82c2-cf07dba06ccd.png)

**Comment by infa-mhadiman** [comment]
Hi @jasonodonnell ,
Sorry, for the delay. I was able to try this with vault-k8s 0.10.0 and vault 1.7 and the vault auth still failed.

Following is the error when using OIDC.
2021-04-29T06:47:59.668Z [WARN]  error assuming role: roleARN=<ROLE> tokenPath=/var/run/secrets/eks.amazonaws.com/serviceaccount/token sessionName= err="WebIdentityErr: failed to retrieve credentials
caused by: InvalidIdentityToken: No OpenIDConnect provider found in your account for https://oidc.eks.us-gov-east-1.amazonaws.com/id/<CLUSTER_ID>
	status code: 400, request id: db7c19dd-09dd-4e83-bdd2-13f6ae173c89"

After this, it tries using the worker node IAM for auth as a fall back which eventually fails.

Some observations:
1. It looks like the vault is using the us-gov-west-1 as the default region in the gov-cloud. our vault setup has the sts_endpoint configured in the auth/aws/config/client to use the sts regional endpoint. In this case, vault should use the regional endpoint instead of the default

2. Please set a default securityContext fsgroup if none is configured for the deployment. This is needed coz if none is configured, the tmpfs volume mount added as part of the mutation will not have read permissions causing a failure again.

Following are the annotations used -
vault.hashicorp.com/agent-inject: "true"
vault.hashicorp.com/auth-path: "auth/aws"
vault.hashicorp.com/auth-type: "aws"
vault.hashicorp.com/role: "VAULT_ROLE"
vault.hashicorp.com/agent-inject-secret-secret1: "SOME_SECRET"
vault.hashicorp.com/auth-config-type: "iam"
vault.hashicorp.com/auth-config-region: "us-gov-east-1"
vault.hashicorp.com/agent-set-security-context: "false"

**Comment by jasonodonnell** [comment]
@infa-mhadiman, can you share the output of your pods description?

`kubectl describe pod <name of your pod with agents>`

**Comment by infa-mhadiman** [comment]
Hi @jasonodonnell,  Please find the describe output below -
```
Name:         mybot-75d979d8fb-ztgll
Namespace:    vault-test
Priority:     0
Node:         <REDACTED>
Start Time:   Wed, 12 May 2021 04:43:11 +0000
Labels:       app=mybot
              pod-template-hash=75d979d8fb
Annotations:  kubernetes.io/psp: eks.privileged
              vault.hashicorp.com/agent-inject: true
              vault.hashicorp.com/agent-inject-secret-secret1: <REDACTED>
              vault.hashicorp.com/agent-inject-status: injected
              vault.hashicorp.com/agent-set-security-context: false
              vault.hashicorp.com/auth-config-region: us-gov-east-1
              vault.hashicorp.com/auth-config-type: iam
              vault.hashicorp.com/auth-path: auth/aws
              vault.hashicorp.com/auth-type: aws
              vault.hashicorp.com/role: <REDACTED>
Status:       Pending
IP:           <REDACTED>
IPs:
  IP:           <REDACTED>
Controlled By:  ReplicaSet/mybot-75d979d8fb
Init Containers:
  vault-agent-init:
    Container ID:  docker://680d84938ff7d5533e9300e77ad5e60c5a4f4d4b8cf9828fede502fe03f1ed45
    Image:         <REDACTED>/vault:1.7.0
    Image ID:      docker-pullable://<REDACTED>/vault@sha256:61a386a6fcb6cb9fa0aa32c8df5c4a913819a6297733ce3a228fa7aaca489f5f
    Port:          <none>
    Host Port:     <none>
    Command:
      /bin/sh
      -ec
    Args:
      echo ${VAULT_CONFIG?} | base64 -d > /home/vault/config.json && vault agent -config=/home/vault/config.json
    State:          Running
      Started:      Wed, 12 May 2021 04:43:13 +0000
    Ready:          False
    Restart Count:  0
    Limits:
      cpu:     500m
      memory:  128Mi
    Requests:
      cpu:     250m
      memory:  64Mi
    Environment:
      VAULT_LOG_LEVEL:              info
      VAULT_LOG_FORMAT:             standard
      VAULT_CONFIG:                 <REDACTED>
      AWS_WEB_IDENTITY_TOKEN_FILE:  /var/run/secrets/eks.amazonaws.com/serviceaccount/token
      AWS_ROLE_ARN:                 <REDACTED>
    Mounts:
      /home/vault from home-init (rw)
      /var/run/secrets/eks.amazonaws.com/serviceaccount from aws-iam-token (ro)
      /var/run/secrets/kubernetes.io/serviceaccount from oidc-sa-token-mw9nc (ro)
      /vault/secrets from vault-secrets (rw)
Containers:
  mybot:
    Container ID:
    Image:         <REDACTED>
    Image ID:
    Port:          <none>
    Host Port:     <none>
    Command:
      sh
      -c
      tail -f /dev/null
    State:          Waiting
      Reason:       PodInitializing
    Ready:          False
    Restart Count:  0
    Limits:
      cpu:     250m
      memory:  256Mi
    Requests:
      cpu:     250m
      memory:  256Mi
    Environment:
      AWS_DEFAULT_REGION:           us-gov-east-1
      AWS_ROLE_ARN:                 <REDACTED>
      AWS_WEB_IDENTITY_TOKEN_FILE:  /var/run/secrets/eks.amazonaws.com/serviceaccount/token
    Mounts:
      /var/run/secrets/eks.amazonaws.com/serviceaccount from aws-iam-token (ro)
      /var/run/secrets/kubernetes.io/serviceaccount from oidc-sa-token-mw9nc (ro)
      /vault/secrets from vault-secrets (rw)
  vault-agent:
    Container ID:
    Image:         <REDACTED>/vault:1.7.0
    Image ID:
    Port:          <none>
    Host Port:     <none>
    Command:
      /bin/sh
      -ec
    Args:
      echo ${VAULT_CONFIG?} | base64 -d > /home/vault/config.json && vault agent -config=/home/vault/config.json
    State:          Waiting
      Reason:       PodInitializing
    Ready:          False
    Restart Count:  0
    Limits:
      cpu:     500m
      memory:  128Mi
    Requests:
      cpu:     250m
      memory:  64Mi
    Environment:
      VAULT_LOG_LEVEL:              info
      VAULT_LOG_FORMAT:             standard
      VAULT_CONFIG:                 <REDACTED>
      AWS_ROLE_ARN:                 <REDACTED>
      AWS_WEB_IDENTITY_TOKEN_FILE:  /var/run/secrets/eks.amazonaws.com/serviceaccount/token
    Mounts:
      /home/vault from home-sidecar (rw)
      /var/run/secrets/eks.amazonaws.com/serviceaccount from aws-iam-token (ro)
      /var/run/secrets/kubernetes.io/serviceaccount from oidc-sa-token-mw9nc (ro)
      /vault/secrets from vault-secrets (rw)
Conditions:
  Type              Status
  Initialized       False
  Ready             False
  ContainersReady   False
  PodScheduled      True
Volumes:
  aws-iam-token:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  86400
  oidc-sa-token-mw9nc:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  oidc-sa-token-mw9nc
    Optional:    false
  home-init:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:     Memory
    SizeLimit:  <unset>
  home-sidecar:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:     Memory
    SizeLimit:  <unset>
  vault-secrets:
    Type:        EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:      Memory
    SizeLimit:   <unset>
QoS Class:       Burstable
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                 node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type    Reason     Age   From               Message
  ----    ------     ----  ----               -------
  Normal  Scheduled  2m4s  default-scheduler  Successfully assigned vault-test/mybot-75d979d8fb-ztgll to <REDACTED>
  Normal  Pulled     2m3s  kubelet            Container image "<REDACTED>/vault:1.7.0" already present on machine
  Normal  Created    2m3s  kubelet            Created container vault-agent-init
  Normal  Started    2m2s  kubelet            Started container vault-agent-init
```

**Comment by jasonodonnell** [comment]
@infa-mhadiman That's for the describe output, but what I really needed was the value for VAULT_CONFIG. It shouldn't contain any sensitive data, but I'm particularly interested in the auto-auth configurations. If there's any of that output (remove specific values you feel uncomfortable sharing) you can share it would be helpful for us to understand if Agent is being configured correctly.

You can decode this config using base64.

**Comment by infa-mhadiman** [comment]
@jasonodonnell, please find the decoded VAULT_CONFIG.
```
{
  "auto_auth": {
    "method": {
      "type": "aws",
      "mount_path": "auth/aws",
      "config": {
        "region": "us-gov-east-1",
        "role": "<REDACTED>",
        "type": "iam"
      }
    },
    "sink": [
      {
        "type": "file",
        "config": {
          "path": "/home/vault/.vault-token"
        }
      }
    ]
  },
  "exit_after_auth": false,
  "pid_file": "/home/vault/.pid",
  "vault": {
    "address": <REDACTED>
  },
  "template": [
    {
      "destination": "/vault/secrets/secret1",
      "contents": "{{ with secret \"REDACTED\" }}{{ range $k, $v := .Data }}{{ $k }}: {{ $v }}\n{{ end }}{{ end }}",
      "left_delimiter": "{{",
      "right_delimiter": "}}"
    }
  ]
}
```

**Comment by jasonodonnell** [comment]
Thanks @infa-mhadiman, that looks correct so I'm unsure why Vault isn't honoring the config. Trying to reproduce on my end.

Also, I wanted to mention that we have an annotation to configure the group of the injected agent: 
`vault.hashicorp.com/agent-run-as-group: 3000`

**Comment by infa-mhadiman** [comment]
@jasonodonnell , I think the problem is with the sts call ? the below error is seen when performing a assumeRoleWithWebIdentity. Looks like it's looking for the OIDC provider url in a different region (us-west-gov-1 which is the default, instead of east-1 ?)
```
Following is the error when using OIDC.
2021-04-29T06:47:59.668Z [WARN] error assuming role: roleARN= tokenPath=/var/run/secrets/eks.amazonaws.com/serviceaccount/token sessionName= err="WebIdentityErr: failed to retrieve credentials
caused by: InvalidIdentityToken: No OpenIDConnect provider found in your account for https://oidc.eks.us-gov-east-1.amazonaws.com/id/<CLUSTER_ID>
status code: 400, request id: db7c19dd-09dd-4e83-bdd2-13f6ae173c89"
```

**Comment by jasonodonnell** [comment]
> Looks like it's looking for the OIDC provider url in a different region (us-west-gov-1 which is the default, instead of east-1 ?)

Sorry for my ignorance, but it appears its using `us-gov-east-1`:

> No OpenIDConnect provider found in your account for https://oidc.eks.us-gov-east-1.amazonaws.com/id/<CLUSTER_ID>

**Comment by jasonodonnell** [comment]
My colleague @tvoran may have found a bug in Vault K8s. When we detect AWS auth method, we copy some environment variables related to IRSA. Default region is missing from this copy: https://github.com/hashicorp/vault-k8s/blob/87fd9ba1348ef1a6d4d649007885cba3319d5c75/agent-inject/agent/agent.go#L629-L643.

**Comment by infa-mhadiman** [comment]
hmm, okay. I was able to quickly verify this. The vault-k8s container does not have a AWS_REGION env variable. so, above would not be possible.

if this env is set for the vault containers injected, this auth goes through fine.

**Comment by tvoran** [comment]
Fixed in #268
